python/pdo/contract/replication.py

```diff
@@ -0,0 +1,214 @@
+            
+        # identify if replication can be skipped, including the case where there is only one provisioned enclave
+        if len(self.storage_clients_to_replicate) == 0:
+            logger.info('Skipping replication: Only one provisioned enclave, so nothing to replicate')
+            return
+       
+        if len(self.blocks_to_replicate) == 0:
+            logger.info('Skipping replication: No change set, so nothing to replicate')
+            return
+
+        # submit individual replication tasks to various storage services asynchronously. Each replica is handled by a separate thread 
+        executor = concurrent.futures.ThreadPoolExecutor(max_workers = len(self.storage_clients_to_replicate))
+        futures_to_replication_tasks = dict() # to keep track of the tasks
+        expiration = self.availability_duration
+        for index, client in enumerate(self.storage_clients_to_replicate):
+            # get data generator. Optimization: If  this generator can be wrapped to be threadsafe, itertools.tee can be used to create
+            # copies without having to read the cache multiple times. Directly using itertools.tee will not work with multi-threadeding
+            block_data_list = ContractState.block_data_generator(self.contract_id, self.blocks_to_replicate, self.data_dir)
+            future = executor.submit(self.push_blocks_to_storage_service, client, block_data_list, expiration)
+            futures_to_replication_tasks[future] = client.ServiceURL
+       
+        #poll for completed replication tasks, wait until num_provable_replicas number of tasks finish
+        urls_of_successful_storage_services = [self.sservice_url_for_contract] #successful by default
+        num_successful_replicas = 1
+        num_unsuccessful_replicas = 0
+        max_allowable_num_unsucessful_replicas = len(self.storage_clients_to_replicate) - self.num_provable_replicas + 1
+            # Add 1 since storage_clients_to_replicate does not contain the sservice corresponding to the contract enclave
+        
+        if num_successful_replicas == self.num_provable_replicas: #covers the case when we only need one proof
+            pass
+        else:
+            for future in concurrent.futures.as_completed(futures_to_replication_tasks):
+                url = futures_to_replication_tasks[future]
+                try:
+                    response = future.result()
+                except Exception as e:
+                    logger.exception('Unexpected situation: Unable to gather sservice response after what looks like successful replication: %s', str(e))
+                    Replicator.exceptions_queue.put('Unexpected situation: Unable to gather sservice response after what looks like successful replication')
+                    raise ReplicationException('Unexpected situation: Unable to gather sservice response after what looks like successful replication') from e
```

Comment:
given that you have three copies of the same string passing to three separate functions and you do more or less the same thing below, might want to think about ways to consolidate these statements